cwm_worker_operator/multiprocessor.py

```diff
@@ -0,0 +1,58 @@
+import time
+import subprocess
+
+
+class Multiprocessor:
+
+    def __init__(self, max_parallel_processes):
+        self.max_parallel_processes = max_parallel_processes
+        self.is_async = max_parallel_processes > 1
+        self.processes = {}
+
+    def _run_async(self, *args, **kwargs):
+        raise NotImplementedError()
+
+    def _run_sync(self,*args, **kwargs):
+        raise NotImplementedError()
+
+    def _get_process_key(self, *args, **kwargs):
+        raise NotImplementedError()
+
+    def get_active_processes(self):
+        process: subprocess.Popen
+        return sum((1 for process in self.processes.values() if process.poll() is None))
+
+    def wait_for_available_process_slot(self):
+        while self.get_active_processes() >= self.max_parallel_processes:
+            time.sleep(1)
+
+    def wait_for_all_processes_to_complete(self):
+        while self.get_active_processes() != 0:
+            time.sleep(1)
+
+    def process(self, *args, **kwargs):
+        if self.is_async:
+            key = self._get_process_key(*args, **kwargs)
+            assert key not in self.processes
+            self.processes[key] = self._run_async(*args, **kwargs)
```

Comment:
How can `_run_async` here enforce that it should always return a process?

Reply:
could check that it's a Popen class using `isinstance` but I prefer not to have too many checks and rely on it raising an Exception if something fails..